tka/aum.go

```diff
@@ -0,0 +1,256 @@
+	Signatures []Signature `cbor:"23,keyasint,omitempty"`
+}
+
+// StaticValidate returns a nil error if the AUM is well-formed.
+func (a *AUM) StaticValidate() error {
+	if a.Key != nil {
+		if err := a.Key.StaticValidate(); err != nil {
+			return err
+		}
+	}
+	if a.PrevAUMHash != nil && len(a.PrevAUMHash) == 0 {
+		return errors.New("absent parent must be represented by a nil slice")
+	}
+	for i, sig := range a.Signatures {
+		if len(sig.KeyID) == 0 || len(sig.Signature) != ed25519.SignatureSize {
+			return fmt.Errorf("signature %d has missing keyID or malformed signature", i)
+		}
+	}
+
+	// TODO(tom): Validate State once a future PR brings in that type.
+
+	switch a.MessageKind {
+	case AUMAddKey:
+		if a.Key == nil {
+			return errors.New("AddKey AUMs must contain a key")
+		}
+		if a.KeyID != nil || a.DisablementSecret != nil || a.State != nil || a.Votes != nil || a.Meta != nil {
+			return errors.New("AddKey AUMs may only specify a Key")
+		}
+	case AUMRemoveKey:
+		if len(a.KeyID) == 0 {
+			return errors.New("RemoveKey AUMs must specify a key ID")
+		}
+		if a.Key != nil || a.DisablementSecret != nil || a.State != nil || a.Votes != nil || a.Meta != nil {
+			return errors.New("RemoveKey AUMs may only specify a KeyID")
+		}
+	case AUMUpdateKey:
+		if len(a.KeyID) == 0 {
+			return errors.New("UpdateKey AUMs must specify a key ID")
+		}
+		if a.Meta == nil && a.Votes == nil {
+			return errors.New("UpdateKey AUMs must contain an update to votes or key metadata")
+		}
+		if a.Key != nil || a.DisablementSecret != nil || a.State != nil {
+			return errors.New("UpdateKey AUMs may only specify KeyID, Votes, and Meta")
+		}
+	case AUMCheckpoint:
+		if a.State == nil {
+			return errors.New("Checkpoint AUMs must specify the state")
+		}
+		if a.KeyID != nil || a.DisablementSecret != nil || a.Key != nil || a.Votes != nil || a.Meta != nil {
+			return errors.New("Checkpoint AUMs may only specify State")
+		}
+	case AUMDisableNL:
+		if len(a.DisablementSecret) == 0 {
+			return errors.New("DisableNL AUMs must specify a disablement secret")
+		}
+		if a.KeyID != nil || a.State != nil || a.Key != nil || a.Votes != nil || a.Meta != nil {
+			return errors.New("DisableNL AUMs may only a disablement secret")
+		}
+	}
+
+	return nil
+}
+
+// Serialize returns the given AUM in a serialized format.
+func (a *AUM) Serialize() []byte {
+	// Why CBOR and not something like JSON?
+	//
+	// The main function of an AUM is to carry signed data. Signatures are
+	// over digests, so the serialized representation must be deterministic.
+	// Further, experience with other attempts (JWS/JWT,SAML,X509 etc) has
+	// taught us that even subtle behaviors such as how you handle invalid
+	// or unrecognized fields + any invariants in subsequent re-serialization
+	// can easily lead to security-relevant logic bugs. Its certainly possible
+	// to invent a workable scheme by massaging a JSON parsing library, though
+	// profoundly unwise.
```

Comment:
I suspect that it's not so much determinism that you need, but rather stability or canonicalization. Determinism is usually defined as the same program producing the same output, but there's no guarantee that different versions of the program or different implementations output the same thing.

I suspect that the property you want is that the serialization is stable where different versions of the program produce the same output. Canonical output goes a step even further than all implementations must produce the same output.

In order of strictness, it goes from non-deterministic to deterministic to stable to canonical.

---

There's been advances in canonical JSON in the recent years. [RFC 8785](https://datatracker.ietf.org/doc/html/rfc8785) defines a canonical representation of JSON, with a Go implementation in [`json.RawValue.Canonicalize`](https://pkg.go.dev/github.com/go-json-experiment/json#RawValue.Canonicalize).

Reply:
Didn't know the nuance of all those terms! In that case, we want 'canonical'.

Between the poor efficiency of encoding raw bytes and its use for signed data being an antipattern (see JWS/JWTs), I think we would have more issues from canonical JSON than we would from something like CBOR.